node/pkg/watchers/sui/watcher.go

```diff
@@ -503,25 +440,208 @@ func (e *Watcher) Run(ctx context.Context) error {
 
 	select {
 	case <-ctx.Done():
-		_ = ws.Close(websocket.StatusNormalClosure, "")
 		return ctx.Err()
 	case err := <-errC:
-		_ = ws.Close(websocket.StatusInternalError, err.Error())
 		return err
 	}
 }
 
-// fixSuiWsURL ensures the websocket scheme is properly specified
-func (e *Watcher) fixSuiWsURL(logger *zap.Logger) error {
-	u, _ := url.Parse(e.suiWS)
+func (w *Watcher) getEvents() ([]SuiResultInfo, error) {
+	// Only get events newer than the last processed height
+	var retVal []SuiResultInfo
+	var results []SuiResult
+	var txs []string
+	var nextCursor struct {
+		TxDigest string
+		EventSeq string
+	}
+	firstTime := true
+	done := false
+	for !done {
+		var payload string
+		if firstTime {
+			payload = w.queryEventsCmd
+		} else {
+			payload = fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, { "txDigest": "%s", "eventSeq": "%s" }, %d, %t]}`,
+				w.suiMoveEventType, nextCursor.TxDigest, nextCursor.EventSeq, w.maximumBatchSize, w.descendingOrder)
+		}
+		res, err := w.suiQueryEvents(payload)
+		if err != nil {
+			return retVal, err
+		}
+		for _, datum := range res.Result.Data {
+			txs = append(txs, *datum.ID.TxDigest)
+			results = append(results, datum)
+		}
+		if (len(res.Result.Data) == 0) || (len(txs) == 0) {
+			// In devnet (tilt) the core contract may not have any events and we don't want to flood the logs.
+			if w.unsafeDevMode {
+				return retVal, nil
+			}
+			return retVal, errors.New("getEvents was unable to get any events")
+		}
+		// Get and check the checkpoint for the last event against the lastProcessedHeight to see if we are done.
+		height, hErr := w.getCheckpointForDigest(txs[len(txs)-1])
+		if hErr != nil {
+			return retVal, hErr
+		}
+		if height <= w.latestProcessedCheckpoint {
+			break
+		}
+		if (!done) && res.Result.HasNextPage {
```

Comment:
I think `done` is always `false` here because this is a `for <bool>` loop.
With that in mind, I think this section could be simplified to:

```go
    ...
    if !res.Result.Has.NextPage {
        break
    }
    nextCursor.TxDigest = res.Result.NextCursor.TxDigest
    nextCursor.EventSeq = res.Result.NextCursor.EventSeq
}
```
In that case you could also change the loop to `for { ... }` instead of `for !done {...}`. 

IMO this style makes reasoning about control flow a little easier. (Compared with tracking a mix of `break`s, `return`s, and checks against `done`.)

Anyway it's a style thing at the end of the day so feel free to ignore.

Reply:
Sounds reasonable.  Changing.